### ISSUE TITLE
Parallel execution cases with deeper suites

### ISSUE THREAD

**Issue by 1999** [yandex/mocha-parallel-tests#110]
Copied from #106 (original comment by @pedgarcia)

@1999 I hope you still see this since it is closed... What happens when you have nested describes in terms of parallelization?

Imagine the following scenario:
```
describe('module', () => {
  describe('submodule A', () => {
    describe('feature B', () => { ... })
         it(A..)
         it(B..)
  });
  it(C...)
  it(D...)
})
```
Will A B C and D run in parallel, or A and B will, then C and D will, but only after A and B?

If you have better examples, please point me at them

Thanks
Paulo

**Comment by 1999** [comment]
@pedgarcia that's an interesting case since when `mocha-parallel-tests` has been started we didn't think of this case. Instead I was thinking only of the simpliest cases like one describe() per file or couple of them on the same level of depth. Only these cases are covered with tests and I can't even predict what's the mocha-parallel-tests behaviour for deeper cases.

Let's try to think on how this should be done in terms of parallel execution. We can run two parallel suites: ("featureB") and ("module" without "submodule A"). However there's often some code in between, for instance there can be a function declared right before "submodule A" and it should be accessible by both of our parallel execution tests. Also it often happens that developers use some variables to store data between the tests. This will definitely fail here because this variable can no longer serve as a data store and the execution will be parallel.

What's your opinion here?

**Comment by pedgarcia** [comment]
I guess the whole issue is that I use `describes` not only to organize the suites but also as scope for variables and/or before/after hooks. My expectation on the scenario I gave is to have all its running in parallel....

But, I think it is a good feature to have same-level describes to be executed in series but not in parallel, like mocha-parallel-tests does today, so the only thing I can think of is an option where you simply ignore that.

For example, by default `mocha-parallel-tests` keeps doing like it does today. If some specific option is used, it could ignore the describes, running all `it` in parallel. I didn't check how this module is implemented or if that idea is doable, though.

On the other hand, I am re-evaluating my test structure to see if I could live with one since level of describe, for example....

What do you think?

**Comment by pedgarcia** [comment]
@1999 I think the following is a more realistic example on how I write the tests:

```
  describe('Test suite #1', function () {

      describe('Test suite #1A', function () {
        it('#1A-1 should end in 5 second', function (done) {
            setTimeout(done, 5000);
        });

        it('#1A-2 should end in 5 second', function (done) {
          setTimeout(done, 5000);
        });
      });

      describe('Test suite #1B', function () {
        it('#1B-1 should end in 5 second', function (done) {
            setTimeout(done, 5000);
        });

        it('#1B-2 should end in 5 second', function (done) {
          setTimeout(done, 5000);
        });
      });

    it('#1-1 should end in 5 second', function (done) {
      setTimeout(done, 5000);
    });

    it('#1-2 should end in 5 second', function (done) {
      setTimeout(done, 5000);
    });

  })
```
So with this example above, I can see there is no parallel execution at all, as the full test takes 30 secs.

Maybe a clear behaviour (if possible) would have `Test suite #1A` and `Test suite #1B` running in parallel as they are at the same level. Also, the suites #2 and #3 (at root level) could be running.

It is still not clear in my mind, but maybe this scenario is possible: (??)

- Root describes are initiated in parallel
- On each root describe, the inner-most describes that are in the same depth would run in parallel
- As they are completed, the execution would go up one level (towards root), executing all describes at that level in parallel
- The last step would run `its` in the root level.

With this, there is still the need to think about functions or variables declared, etc...

How to do all this is above my head :) and maybe the current architecture wouldn't permit such implementation. It would be nice to have more people chiming in....

**Comment by 1999** [comment]
> So today mocha-parallel-tests runs describes in parallel, but the its inside them in series, correct?

correct

**Comment by 1999** [comment]
cc @k03mad @kolontsov @mmotkina guys please take a look at discussion here and tell what you think.

**Comment by k03mad** [comment]
I think that parallelization inside one file is not good way at all.
If test so capacious, that it need to run it parts in parallel, such parts should be in separate independent files.

This seems to be less confusing, less flaky and complies with the best practices "small + atomic + autonomous".

**Comment by 1999** [comment]
@k03mad that also makes sense.
@pedgarcia what do you think of this approach? Like "if you want to run your tests in parallel, adjust your test files to contain simplier code"

**Comment by pedgarcia** [comment]
@1999, I spent more time thinking about my suggestion and reviewing my test code and I have to agree that it would turn everything way to complicated. I agree the approach to simplify the tests makes more sense....

**Comment by 1999** [comment]
UPD: The final idea is to run all files + all **flat** suites in these files in parallel.

**Comment by k03mad** [comment]
Will it be optional?
I think it will break many mpt-users tests (not only kp:))

**Comment by 1999** [comment]
This will be a new major version of mocha-parallel-tests, not an option.
